File: trees/bst/serialize_and_deserialize_bst.rb
# Definition for a binary tree node.
class TreeNode
    attr_accessor :val, :left, :right

    def initialize(val)
        @val = val
        @left, @right = nil, nil
    end
end

class Integer
    N_BYTES = [42].pack('i').size
    N_BITS = N_BYTES*8
    MAX = 2**(N_BITS - 2) - 1
    MIN = -MAX - 1
end

# Encodes a tree to a single string.
# @param {TreeNode} root
# @return {string}
def serialize(root)
    pre_order = []
    dfs_serialize(root, pre_order)
    pre_order.join(",")
end

def dfs_serialize(node, result)
    return if !node
    result.push(node.val.to_s)
    dfs_serialize(node.left, result)
    dfs_serialize(node.right, result)
end

# Decodes your encoded data to tree.
# @param {string} data
# @return {TreeNode}
def deserialize(data)
    return if data.empty?
    v_min, v_max = Integer::MIN, Integer::MAX
    pre_order = data.split(",")
    dfs_deserialize(pre_order, v_min, v_max)
end

def dfs_deserialize(pre_order, v_min, v_max)
    return if pre_order.empty?
    val = pre_order.first.to_i
    return if !val.between?(v_min, v_max)

    pre_order.shift()
    node = TreeNode.new(val)
    node.left = dfs_deserialize(pre_order, v_min, val)
    node.right = dfs_deserialize(pre_order, val, v_max)

    node
end

# 449. Serialize and Deserialize BST
# https://leetcode.com/problems/serialize-and-deserialize-bst/

# Key Insight
# Difference from from problem no. 297 is:
# 1. This is a BST
# 2. We need to make the serialized string as compact as possible
# 3. All we need to do is get the order of insertion into BST
#    to construct the bst,
#    we don't need to know position of leaf nodes and such and
#    hence it's as compact as possible.

# Approach 1: DFS using pre-order traversal
# Steps:
# 1. Serialize using pre-order dfs
# 2. Deserialize using pre-order traversal of array and 
#    use node bounds to find the leaves

# Time: O(n)
# Space: O(n)

require 'test/unit'
extend Test::Unit::Assertions

#       6
#      / \
#     4   8
#    / \
#   2   5

root = TreeNode.new(6)
l, r = TreeNode.new(4), TreeNode.new(8)
l1, l2 = TreeNode.new(2), TreeNode.new(5)
root.left, root.right = l, r
l.left, l.right = l1, l2

assert_equal(serialize(root), "6,4,2,5,8")

new_root = deserialize("6,4,2,5,8")
assert_equal(new_root.val, root.val)
assert_equal(new_root.left.val, l.val)
assert_equal(new_root.right.val, r.val)
assert_equal(new_root.left.left.val, l1.val)
assert_equal(new_root.left.right.val, l2.val)

assert_equal(serialize(nil), "")
assert_equal(deserialize(""), nil)

File: trees/bst/serialize_and_deserialize_bst.py
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Codec:
    def serialize(self, root):
        self.pre_order = []
        self.dfs_serialize(root)
        return ','.join(self.pre_order)

    def dfs_serialize(self, node):
        if not node: return

        self.pre_order.append(str(node.val))
        self.dfs_serialize(node.left)
        self.dfs_serialize(node.right)

    def deserialize(self, data):
        if not data: return

        pre_order = data.split(',')
        return self.dfs_deserialize(pre_order)

    def dfs_deserialize(self, pre_order, vmin=float('-inf'), vmax=float('inf')):
        if not pre_order: return
        val = int(pre_order[0])
        if not vmin <= val <= vmax: return
        pre_order.pop(0)

        root = TreeNode(val)
        root.left = self.dfs_deserialize(pre_order, vmin, val - 1)
        root.right = self.dfs_deserialize(pre_order, val + 1, vmax)
        return root


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))

# 449. Serialize and Deserialize BST
# https://leetcode.com/problems/serialize-and-deserialize-bst/

# Complexity Analysis
# Time complexity: in both serialization and deserialization functions,
#                  we visit each node exactly once, thus the time complexity is O(N),
#                  where N is the number of nodes, i.e. the size of tree.
# Space complexity: in both serialization and deserialization functions,
#                   we keep the entire tree, either at the beginning or at the end,
#                   therefore, the space complexity is O(N).

# Time: O(n)
# Space: O(n)

# Diff from #297.
# Like others have noted, the primary difference in the problem statement is:
# the encoded string needs to be as compact as possible.

# One of the ways a BST tree is different from a general binary tree is:
# its structure is wholly dependent on the order in which the values are inserted.
# A string created from a preorder traversal of a BST will tell you the order in which
# the values were inserted into the tree. Since you just need the order the values were inserted,
# you do not need to account for null nodes in the string with "#" or "null".
# Hence, the final string contains only the values and separators, which makes it the most compact possible.

sol = Codec()
print(sol.serialize(None))
